panels/updates/cc-updates-panel.c

```diff
@@ -0,0 +1,684 @@
+  GtkWidget          *network_status_icon;
+  GtkWidget          *scheduled_updates_switch;
+  CcTariffEditor     *tariff_editor;
+
+  /* Network Manager */
+  NMClient           *nm_client;
+  NMDevice           *current_device;
+
+  /* Signal handlers */
+  guint               changed_id;
+  guint               save_tariff_timeout_id;
+
+  GCancellable       *cancellable;
+};
+
+
+static void          on_automatic_updates_switch_changed_cb      (GtkSwitch      *sw,
+                                                                  GParamSpec     *pspec,
+                                                                  CcUpdatesPanel *self);
+
+static void          on_network_changed_cb                       (CcUpdatesPanel *self);
+
+static void          on_network_changes_commited_cb              (GObject        *source,
+                                                                  GAsyncResult   *result,
+                                                                  gpointer        user_data);
+
+static void          on_scheduled_updates_switch_changed_cb      (GtkSwitch      *sw,
+                                                                  GParamSpec     *pspec,
+                                                                  CcUpdatesPanel *self);
+
+static void          on_tariff_changed_cb                        (CcTariffEditor *tariff_editor,
+                                                                  CcUpdatesPanel *self);
+
+static void          on_tariff_file_deleted_cb                   (GFile          *file,
+                                                                  GAsyncResult   *result,
+                                                                  CcUpdatesPanel *self);
+
+
+G_DEFINE_TYPE (CcUpdatesPanel, cc_updates_panel, CC_TYPE_PANEL)
+
+enum
+{
+  PROP_PARAMETERS = 1,
+  N_PROPS
+};
+
+
+/*
+ * Auxiliary methods
+ */
+
+static void
+set_automatic_updates_setting (NMConnection   *connection,
+                               gboolean        enabled)
+{
+  NMSettingUser *setting_user;
+  g_autoptr(GError) error = NULL;
+
+  g_debug ("Setting "NM_SETTING_ALLOW_DOWNLOADS_WHEN_METERED" to %d", enabled);
+
+  setting_user = NM_SETTING_USER (nm_connection_get_setting (connection, NM_TYPE_SETTING_USER));
+
+  nm_setting_user_set_data (setting_user,
+                            NM_SETTING_ALLOW_DOWNLOADS_WHEN_METERED,
+                            enabled ? "1" : "0",
+                            &error);
+
+  if (error)
+    {
+      g_warning ("Error storing "NM_SETTING_ALLOW_DOWNLOADS_WHEN_METERED": %s", error->message);
+      return;
+    }
+
+  nm_remote_connection_commit_changes_async (NM_REMOTE_CONNECTION (connection),
+                                             TRUE, /* save to disk */
+                                             NULL,
+                                             on_network_changes_commited_cb,
+                                             NULL);
```

Comment:
This async call doesn’t hold a reference to `self`, so it’s possible that the `CcUpdatesPanel` will be destroyed before the async operation finishes. This is a bit tricky to fix. There are two possible solutions:
 * Hold a strong reference to the `CcUpdatesPanel`, which means that a slow async operation can keep the panel alive longer than expected.
 * Hold a weak reference, and pass in a `GCancellable` to the async call. When the `CcUpdatesPanel` is disposed, cancel the `GCancellable` so that the async operation completes.

Reply:
The callback does not access anything from `CcUpdatesPanel`. I don't understand why it is necessary to keep the panel alive in this case.

Reply:
If, for example, the user changes the tariff, then immediately closes g-c-c, the following could happen:
 * Saving the changes starts
 * The async start function returns, and control goes back to the main context
 * The main loop is quit (I suspect that closing the g-c-c window will end up calling `g_main_loop_quit()` to achieve this, but haven’t checked)
 * The necessary main context iterations to finish the D-Bus communication necessary to finish saving the NetworkManager settings do not happen, and hence the settings potentially don’t get saved

In all likelihood, the D-Bus traffic here is a single message, and so the settings will be correctly saved, even if the user does somehow manage to hit the race described above. However, proving that the savings will always be saved properly is really hard.

What needs to happen is that the `GMainContext` continues to be iterated on shutdown of g-c-c until all pending operations (from all panels) are compete (after hiding the main window). However, I realise that’s a fairly major change.

For the moment, I’d be fine with a `FIXME` comment here mentioning that, and a plan to implement this blocking shutdown at the same time as making panel startup async (because that will expose more of the same kind of race).